### DISCUSSION TITLE
81. 클러스터형 인덱스와 논클러스터형 인덱스 테이블 구조의 차이를 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#83] [General]
클러스터형 인덱스와 논클러스터형 인덱스 테이블 구조의 차이를 설명해 주세요.

**Comment by linirini** [comment]
1. 클러스터형 인덱스(Clustered Index)
- 테이블 전체가 정렬된 인덱스가 되는 방식(테이블 레코드들이, 인덱스 컬럼의 정렬 순서대로 적재되어 있는 것)
- 테이블 데이터는 클러스터형 인덱스의 정렬 순서에 따라 디스크에 물리적으로 정렬된다.
- 테이블 당 하나민 생성할 수 있다.
- InnoDB에서는 Primary Key가 클러스터형 인덱스로 설정된다.
- Primary Key가 없을 경우, 내부적으로 Unique Key를 클러스터형 인덱스로 사용하거나, 내부 Row ID를 생성하여 클러스터형 인덱스를 만든다.
- B+Tree 형태로 구성되어 있다.
- Insert 시 Data 가 정렬되고 Index 는 Data Block 의 첫 번째 레코드의 주소값을 가지게 되며, Index 가 곧 바로 Data Block 에 접근해서 Secondary Index(Non-Clustered Index) 보다 보통 동작이 빠르다.
- Data 가 정렬되어 저장되므로, Secondary Index 에 비해 범위로 질의를 하는 것에 유리하다.(CUD에 대해서는 더 느리다.)
- 인덱스 자체의 리프 페이지가 데이터 -> 인덱스 자체에 데이터가 포함

2. 논클러스터형 인덱스(Non-Clustered Index)
- 인덱스와 데이터가 별도로 저장
- 클러스터형 인덱스와 다르게 물리적으로 테이블을 정렬하지 않는다.
- 정렬된 별도의 인덱스 페이지를 생성하고 관리한다.
- 인덱스는 특정 칼럼에 대한 검색 최적화를 제공하며, 리프 노드에는 데이터의 참조 정보가 저장된다.
- 데이터 참조 정보란 InnoDB에서는 클러스터형 인덱스의 키 값에 해당한다.
- 이를 통해 논클러스터형 인덱스를 검색한 후,  PK를 이용하여 실제 데이터에 접근한다 : Back To Table 과정
